File: fix_posts_final.py
import re
import subprocess
import sys

def extract_posts_array(content):
    """Return the inner content of the posts array (between '[' and ']')"""
    start = content.find('const posts = [')
    if start == -1:
        raise ValueError('Could not find const posts = [')
    start += len('const posts = [')
    end = content.rfind('];')
    if end == -1:
        raise ValueError('Could not find ];')
    inner = content[start:end].strip()
    return inner

def split_posts(inner):
    """Split the inner string into a list of post objects (as strings including braces)."""
    posts = []
    brace_count = 0
    in_string = False
    escape = False
    current = []
    i = 0
    while i < len(inner):
        c = inner[i]
        if escape:
            escape = False
            current.append(c)
            i += 1
            continue
        if c == '\\\\':
            escape = True
            current.append(c)
            i += 1
            continue
        if c == '"' and not in_string:
            in_string = not in_string
            current.append(c)
            i += 1
            continue
        if not in_string:
            if c == '{':
                brace_count += 1
                current.append(c)
                i += 1
                continue
            if c == '}':
                brace_count -= 1
                current.append(c)
                i += 1
                if brace_count == 0:
                    posts.append(''.join(current))
                    current = []
                    continue
        else:
            current.append(c)
            i += 1
            continue
    # If there is leftover (should not happen)
    if current:
        posts.append(''.join(current))
    return posts

def fix_excerpt_in_post(post_obj):
    """Given a post object string, ensure its excerpt field is a proper double-quoted string with escaped newlines."""
    # Find excerpt:
    excerpt_pos = post_obj.find('excerpt:')
    if excerpt_pos == -1:
        return post_obj
    # Find start of string literal after excerpt:
    j = excerpt_pos + len('excerpt:')
    while j < len(post_obj) and post_obj[j] in ' \\t\\n\\r':
        j += 1
    if j >= len(post_obj):
        return post_obj
    opening = post_obj[j]
    if opening not in ('"', '`'):
        return post_obj
    # Find matching closing quote
    k = j + 1
    escaped = False
    while k < len(post_obj):
        c = post_obj[k]
        if escaped:
            escaped = False
            k += 1
            continue
        if c == '\\\\':
            escaped = True
            k += 1
            continue
        if c == opening and not escaped:
            k += 1  # include closing quote
            break
        k += 1
    else:
        return post_obj  # no closing quote found
    # Extract inner content
    if opening == '`':
        inner = post_obj[j+1:k-1]
    else:  # opening == '"'
        inner = post_obj[j+1:k-1]
    # Fix inner: replace newline with \\n, escape backslashes and double quotes
    fixed = inner.replace('\\\\', '\\\\\\\\').replace('"', '\\\\"').replace('\\n', '\\\\n').replace('\\r', '\\\\r')
    # Build new literal
    new_literal = '"' + fixed + '"'
    # Replace
    new_post = post_obj[:j] + new_literal + post_obj[k:]
    return new_post

def main():
    slug_target = "russian-oil-tanker-reaches-cuba-after-trump-appears-to-loosen-blockade"
    # Get current posts.js
    with open('js/posts.js', 'r') as f:
        current_content = f.read()
    # Get old posts.js from commit 8dc6518
    try:
        old_content = subprocess.check_output(['git', 'show', '8dc6518:js/posts.js'], text=True, stderr=subprocess.DEVNULL)
    except subprocess.CalledProcessError as e:
        print(f"Error retrieving old posts.js: {e}")
        sys.exit(1)
    # Extract arrays
    try:
        current_inner = extract_posts_array(current_content)
        old_inner = extract_posts_array(old_content)
    except ValueError as e:
        print(f"Error extracting posts array: {e}")
        sys.exit(1)
    # Split into posts
    current_posts = split_posts(current_inner)
    old_posts = split_posts(old_inner)
    # Find the target post in old_posts
    target_post_obj = None
    for post in old_posts:
        if f'slug: "{slug_target}"' in post or f"slug: '{slug_target}'" in post:
            target_post_obj = post
            break
    if target_post_obj is None:
        print(f"Error: Could not find post with slug {slug_target} in old posts")
        sys.exit(1)
    # Fix the excerpt in the target post
    target_post_fixed = fix_excerpt_in_post(target_post_obj)
    # Ensure it ends with a comma (since we will insert it between two posts)
    target_post_fixed = target_post_fixed.rstrip()
    if not target_post_fixed.endswith(','):
        target_post_fixed = target_post_fixed + ','
    # Now we want to construct new posts list:
    # We keep the first post from current_posts (the newly generated post)
    # Then we insert the target_post_fixed
    # Then we append the rest of the old_posts, but we need to skip the target post we already used, and also skip any posts that are already in current_posts? 
    # However, we want to preserve the order as it was in the old array, but with the new post added at the front.
    # The current_posts currently contains: [new_post, ...] where the ... is the old array but with the russian post missing? Actually we need to check.
    # Let's instead construct the new array as:
    #   new_post (from current_posts[0])
    #   target_post_fixed
    #   then all posts from old_posts except the target_post, in their original order.
    # But we also need to ensure we don't duplicate any other posts that might have been added after the old commit (there shouldn't be any).
    # So we will take old_posts, remove the target_post, and then append them.
    # However, we also need to consider that the current_posts[0] might be a duplicate of a post already in old_posts? It shouldn't be because it's newly generated.
    # Let's get the slug of the first post in current_posts to confirm.
    first_post = current_posts[0] if current_posts else None
    if first_post:
        # extract slug
        slug_match = re.search(r'slug\\s*:\\s*["\\']([^"\\']+)["\\']', first_post)
        if slug_match:
            first_slug = slug_match.group(1)
            print(f"First post slug: {first_slug}")
    # Build list of old posts without the target
    old_posts_without_target = [p for p in old_posts if p != target_post_obj]
    # Now construct new posts list: [first_post, target_post_fixed] + old_posts_without_target
    new_posts_list = []
    if first_post:
        new_posts_list.append(first_post)
    new_posts_list.append(target_post_fixed)
    new_posts_list.extend(old_posts_without_target)
    # Now we need to fix the excerpt in all posts (just in case) but we already fixed the target; we should also fix the first_post and others.
    # We'll apply fix_excerpt_in_post to each post.
    new_posts_list = [fix_excerpt_in_post(p) for p in new_posts_list]
    # Ensure each post except the last ends with a comma
    for i in range(len(new_posts_list)):
        post = new_posts_list[i].rstrip()
        if i != len(new_posts_list) - 1:
            if not post.endswith(','):
                post = post + ','
        new_posts_list[i] = post
    # Join with newlines
    inner_new = '\\n'.join(new_posts_list)
    new_content = 'const posts = [' + inner_new + '];'
    # Replace the const posts = ... line in the current content
    # We'll replace from the start of 'const posts = [' to the closing '];'
    updated_content = re.sub(r'const posts = \\[\\s*[\\s\\S]*?\\];', new_content, current_content, count=1)
    # Write back
    with open('js/posts.js', 'w') as f:
        f.write(updated_content)
    print("Successfully restored russian post and fixed posts.js")

if __name__ == '__main__':
    main()

File: fix_posts_final.js
const fs = require('htmlparser2'); // Not available, we'll do without
// We'll do a simpler approach: just fix the backup2 file and then add the missing posts from original_posts.js? 
// Let's instead: take original_posts.js (clean) and then for each missing slug, we'll get the post from backup2 and fix it.
// We know backup2 has the posts we need (except Starmer and maybe others) but with syntax errors.

// Function to fix a string: remove ** and replace actual newlines in single/double quoted strings with \n
function fixStringInObject(str) {
  let result = '';
  let i = 0;
  let len = str.length;
  let inSingle = false;
  let inDouble = false;
  let inBacktick = false;
  let escape = false;

  while (i < len) {
    let c = str[i];
    if (escape) {
      escape = false;
      result += c;
      i++;
      continue;
    }
    if (c === '\\\\') {
      escape = true;
      result += c;
      i++;
      continue;
    }
    // Handle quotes
    if (!inBacktick) {
      if (c === "'" && !inDouble) {
        inSingle = !inSingle;
        result += c;
        i++;
        continue;
      }
      if (c === '"' && !inSingle) {
        inDouble = !inDouble;
        result += c;
        i++;
        continue;
      }
    }
    if (c === '`' && !(inSingle || inDouble)) {
      inBacktick = !inBacktick;
      result += c;
      i++;
      continue;
    }
    // Replace newlines inside string literals (single or double) with \n
    if (c === '\n' || c === '\r') {
      if (inSingle || inDouble) {
        result += '\\\\n';
      } else {
        result += c;
      }
    } else {
      // Remove markdown bold markers by not adding them
      if (c === '*' && i + 1 < len && str[i + 1] === '*') {
        // Skip both asterisks
        i += 2;
        continue;
      } else {
        result += c;
      }
    }
    i++;
  }
  return result;
}

// Extract posts array content from a file given the file path
function extractPostsArrayContent(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  const match = content.match(/const posts = \[([\s\S]*?)\];/);
  if (!match) {
    throw new Error(`Could not find posts array in ${filePath}`);
  }
  return match[1]; // the content inside the brackets
}

// Split array content into post objects (strings) by top-level braces
function splitPosts(arrayContent) {
  const posts = [];
  let braceCount = 0;
  let current = '';
  for (let i = 0; i < arrayContent.length; i++) {
    const c = arrayContent[i];
    if (c === '{') {
      if (braceCount === 0) {
        current = '{';
      } else {
        current += c;
      }
      braceCount++;
    } else if (c === '}') {
      braceCount--;
      current += c;
      if (braceCount === 0) {
        posts.push(current);
        current = '';
      } else {
        current += c;
      }
    } else {
      if (braceCount > 0) {
        current += c;
      }
    }
  }
  return posts;
}

// Extract slug from a post object string
function extractSlug(postStr) {
  const match = postStr.match(/slug:\s*['\"]([^'\"]+)['\"]/);
  return match ? match[1] : null;
}

// Read original_posts.js (clean base)
let originalArrayContent;
try {
  originalArrayContent = extractPostsArrayContent('original_posts.js');
} catch (e) {
  console.error(e.message);
  process.exit(1);
}
const originalPosts = splitPosts(originalArrayContent);
console.log(`Original posts count: ${originalPosts.length}`);

// Get slugs from original posts
const originalSlugs = originalPosts.map(p => extractSlug(p)).filter(Boolean);
console.log('Original slugs:', originalSlugs.join(','));

// List of all expected slugs from HTML files (we'll compute from the posts directory)
// Exclude article-template.html as it's likely a template, not a post.
const htmlSlugs = [
  'ai-image-generation-chatgpt-claude-grok-2026',
  'bird-flu-kerala-india-h5n1-2026',
  'chuck-norris-dies-at-86',
  'iea-global-economy-major-threat-iran-war-march-2026',
  'iran-us-war-natanz-nuclear-strike-march-2026',
  'nasa-moon-rocket-launch-pad-april-2026',
  'rocket-lab-launches-8th-satellite-synspective',
  'russian-oil-tanker-reaches-cuba-after-trump-appears-to-loosen-blockade',
  'scientists-grow-hair-follicles-lab-breakthrough-2026',
  'spain-closes-airspace-to-us-aircraft-involved-in-iran-war',
  'starmergivesdoctors48hourstocancelstrikeorlosenewjobspackage',
  'strait-of-hormuz-oil-crisis-2026',
  'trump-48-hour-ultimatum-iran-hormuz-march-2026',
  'trump-claims-iran-talks-denied-ultimatum-extended-march-2026',
  'us-deploys-82nd-airborne-ceasefire-talks-iran-war-march-2026',
  'why-does-the-us-have-iran-s-kharg-island-in-its-sights'
];

// Determine missing slugs
const missingSlugs = htmlSlugs.filter(s => !originalSlugs.includes(s));
console.log('Missing slugs:', missingSlugs.join(','));
console.log('Missing count:', missingSlugs.length);

// Backup files to search (in order of preference)
const backupFiles = [
  'js/posts.js.backup2',
  'js/posts.js.backup3',
  'js/posts.js.backup_broken',
  'old_posts.js'
];

// For each missing slug, find the post in the backups
const missingPosts = []; // will hold fixed post strings
for (const slug of missingSlugs) {
  let found = false;
  for (const file of backupFiles) {
    try {
      const arrayContent = extractPostsArrayContent(file);
      const posts = splitPosts(arrayContent);
      for (const postStr of posts) {
        if (extractSlug(postStr) === slug) {
          // Fix the post object string
          const fixed = fixStringInObject(postStr);
          missingPosts.push(fixed);
          console.log(`  Found ${slug} in ${file}`);
          found = true;
          break;
        }
      }
      if (found) break;
    } catch (e) {
      // If the file doesn't have a posts array or other error, skip
      console.error(`Error processing ${file}: ${e.message}`);
    }
  }
  if (!found) {
    console.error(`Could not find post for slug: ${slug}`);
    process.exit(1);
  }
}

// Now construct the new array content: originalPosts (unchanged) + missingPosts
let newArrayContent = '';
// Add original posts (as they are, unchanged)
originalPosts.forEach((postStr, index) => {
  newArrayContent += postStr;
  if (index !== originalPosts.length - 1 || missingPosts.length > 0) {
    newArrayContent += ',\n';
  }
});
// Add fixed missing posts
missingPosts.forEach((postStr, index) => {
  newArrayContent += postStr;
  if (index !== missingPosts.length - 1) {
    newArrayContent += ',\n';
  }
});

// Build the new file content
const newContent = `// posts.js — Article database
// Each post: { slug, title, excerpt, image, category, date, author, url, body }

const posts = [
${newArrayContent}
];`;

fs.writeFileSync('js/posts.js', newContent);
console.log('Written js/posts.js');

// Verify by counting slugs
const test = fs.readFileSync('js/posts.js', 'utf8');
const postsMatch = test.match(/const posts = \[([\s\S]*?)\];/);
if (postsMatch) {
  const slugMatches = postsMatch[1].match(/slug:/g);
  const count = slugMatches ? slugMatches.length : 0;
  console.log(`New posts.js has ${count} posts`);
  // Also list slugs to verify
  const slugList = postsMatch[1].match(/slug:\s*['\"]([^'\"]+)['\"]/g);
  if (slugList) {
    const slugs = slugList.map(m => m.match(/['\"]([^'\"]+)['\"]/)[1]);
    console.log('Slugs:', slugs.join(','));
  }
} else {
  console.error('Could not find posts array in new file');
  process.exit(1);
}
